File: src/lib/userService.ts
import { WithId } from 'mongodb';
import { randomBytes, scryptSync, timingSafeEqual } from 'crypto';
import { getDb } from './db';
import { User, UserRole } from '@/types';

const COLLECTION_NAME = 'usuarios';
const PASSWORD_KEY_LENGTH = 64;
const ALLOWED_ROLES: UserRole[] = ['admin', 'user'];
const MIN_PASSWORD_LENGTH = 8;

type UserDocument = Omit<User, '_id' | 'createdAt' | 'updatedAt'> & {
  email: string;
  role: UserRole;
  passwordHash: string;
  createdAt: Date;
  updatedAt: Date;
};

export interface RegisterPayload {
  email?: string;
  password?: string;
  name?: string;
  role?: string;
}

export interface LoginPayload {
  email?: string;
  password?: string;
}

function mapUser(doc: WithId<UserDocument>): User {
  return {
    _id: doc._id.toHexString(),
    email: doc.email,
    name: doc.name,
    role: doc.role,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

function ensureRole(role?: string): UserRole {
  if (!role) {
    return 'user';
  }
  const normalized = role.trim().toLowerCase();
  if (ALLOWED_ROLES.includes(normalized as UserRole)) {
    return normalized as UserRole;
  }
  throw new Error(`Rol no válido. Roles permitidos: ${ALLOWED_ROLES.join(', ')}.`);
}

function ensurePassword(password?: string): string {
  if (!password || typeof password !== 'string' || password.length < MIN_PASSWORD_LENGTH) {
    throw new Error(`La contraseña debe tener al menos ${MIN_PASSWORD_LENGTH} caracteres.`);
  }
  return password;
}

function hashPassword(password: string): string {
  const salt = randomBytes(16);
  const derivedKey = scryptSync(password, salt, PASSWORD_KEY_LENGTH);
  return `${salt.toString('hex')}:${derivedKey.toString('hex')}`;
}

function verifyPassword(password: string, storedHash: string): boolean {
  const [saltHex, hashHex] = storedHash.split(':');
  if (!saltHex || !hashHex) {
    return false;
  }
  const saltBuffer = Buffer.from(saltHex, 'hex');
  const storedBuffer = Buffer.from(hashHex, 'hex');
  const derived = scryptSync(password, saltBuffer, PASSWORD_KEY_LENGTH);
  if (storedBuffer.length !== derived.length) {
    return false;
  }
  return timingSafeEqual(derived, storedBuffer);
}

export async function registerUser(payload: RegisterPayload): Promise<User> {
  if (!payload.email) {
    throw new Error('El correo es obligatorio.');
  }
  const email = normalizeEmail(payload.email);
  const password = ensurePassword(payload.password);
  const role = ensureRole(payload.role);

  const db = await getDb();
  const collection = db.collection<UserDocument>(COLLECTION_NAME);

  const existing = await collection.findOne({ email });
  if (existing) {
    throw new Error('Ya existe un usuario con ese correo.');
  }

  const now = new Date();
  const document: UserDocument = {
    email,
    name: payload.name?.trim() || undefined,
    role,
    passwordHash: hashPassword(password),
    createdAt: now,
    updatedAt: now,
  };

  const { insertedId } = await collection.insertOne(document);
  const inserted = await collection.findOne({ _id: insertedId });

  if (!inserted) {
    throw new Error('No se pudo registrar el usuario.');
  }

  return mapUser(inserted);
}

export async function authenticateUser(payload: LoginPayload): Promise<User> {
  if (!payload.email || !payload.password) {
    throw new Error('Correo y contraseña son obligatorios.');
  }

  const email = normalizeEmail(payload.email);
  const db = await getDb();
  const collection = db.collection<UserDocument>(COLLECTION_NAME);

  const user = await collection.findOne({ email });
  if (!user) {
    throw new Error('Credenciales inválidas.');
  }

  const validPassword = verifyPassword(payload.password, user.passwordHash);
  if (!validPassword) {
    throw new Error('Credenciales inválidas.');
  }

  return mapUser(user);
}

File: backend/app/services/user_service.py
"""
User registration/authentication service (passwords hashed with scrypt).
"""
from __future__ import annotations

import os
from datetime import datetime
from hashlib import scrypt
from secrets import compare_digest, token_bytes
from typing import Any, Dict, List, Optional

from pymongo.collection import Collection

from app.database import get_database

COLLECTION_NAME = "usuarios"
PASSWORD_KEY_LENGTH = 64
ALLOWED_ROLES = {"admin", "user"}
MIN_PASSWORD_LENGTH = 8


def _collection() -> Collection:
    return get_database()[COLLECTION_NAME]


def _map_user(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "_id": str(doc["_id"]),
        "email": doc["email"],
        "name": doc.get("name"),
        "role": doc["role"],
        "createdAt": doc["createdAt"].isoformat() if doc.get("createdAt") else None,
        "updatedAt": doc["updatedAt"].isoformat() if doc.get("updatedAt") else None,
    }


def _normalize_email(email: str) -> str:
    return email.strip().lower()


def _ensure_role(role: Optional[str]) -> str:
    if not role:
        return "user"
    normalized = role.strip().lower()
    if normalized in ALLOWED_ROLES:
        return normalized
    raise ValueError(f"Rol no válido. Roles permitidos: {', '.join(ALLOWED_ROLES)}.")


def _ensure_password(password: Optional[str]) -> str:
    if not password or len(password) < MIN_PASSWORD_LENGTH:
        raise ValueError(f"La contraseña debe tener al menos {MIN_PASSWORD_LENGTH} caracteres.")
    return password


def _hash_password(password: str) -> str:
    salt = token_bytes(16)
    derived = scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=2**14,
        r=8,
        p=1,
        dklen=PASSWORD_KEY_LENGTH,
    )
    return f"{salt.hex()}:{derived.hex()}"


def _verify_password(password: str, stored_hash: str) -> bool:
    try:
        salt_hex, hash_hex = stored_hash.split(":")
    except ValueError:
        return False
    salt = bytes.fromhex(salt_hex)
    stored = bytes.fromhex(hash_hex)
    derived = scrypt(
        password.encode("utf-8"),
        salt=salt,
        n=2**14,
        r=8,
        p=1,
        dklen=len(stored),
    )
    return compare_digest(stored, derived)


def register_user(payload: Dict[str, Any]) -> Dict[str, Any]:
    if not payload.get("email"):
        raise ValueError("El correo es obligatorio.")
    
    invitation_code = payload.get("invitationCode")
    if not invitation_code:
        raise ValueError("El código de invitación es obligatorio.")
        
    # Validate code and get role
    from app.services.invitation_service import validate_invitation_code, mark_invitation_code_as_used
    role = validate_invitation_code(invitation_code)

    email = _normalize_email(payload["email"])
    password = _ensure_password(payload.get("password"))
    # role is now determined by the invitation code

    if _collection().find_one({"email": email}):
        raise ValueError("Ya existe un usuario con ese correo.")

    now = datetime.utcnow()
    document = {
        "email": email,
        "name": payload.get("name") or None,
        "role": role,
        "passwordHash": _hash_password(password),
        "createdAt": now,
        "updatedAt": now,
    }
    result = _collection().insert_one(document)
    user_id = result.inserted_id
    document["_id"] = user_id
    
    # Mark code as used
    try:
        mark_invitation_code_as_used(invitation_code, str(user_id))
    except Exception as e:
        print(f"Error marking invitation code as used: {e}")
        
    return _map_user(document)


def authenticate_user(payload: Dict[str, Any]) -> Dict[str, Any]:
    if not payload.get("email") or not payload.get("password"):
        raise ValueError("Correo y contraseña son obligatorios.")
    email = _normalize_email(payload["email"])
    user = _collection().find_one({"email": email})
    if not user or not _verify_password(payload["password"], user["passwordHash"]):
        raise ValueError("Credenciales inválidas.")
    return _map_user(user)


def list_users() -> List[Dict[str, Any]]:
    cursor = _collection().find().sort("createdAt", -1)
    return [_map_user(doc) for doc in cursor]


def is_admin(email: str) -> bool:
    user = _collection().find_one({"email": _normalize_email(email)})
    return user is not None and user.get("role") == "admin"
